crates/iceberg/src/writer/mod.rs

```diff
@@ -99,12 +201,12 @@ mod tests {
     use arrow_select::concat::concat_batches;
     use parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder;
 
-    use super::IcebergWriter;
+    use super::{BoxedIcebergWriter, IcebergWriter};
     use crate::io::FileIO;
     use crate::spec::{DataFile, DataFileFormat};
 
     // This function is used to guarantee the trait can be used as a object safe trait.
-    async fn _guarantee_object_safe(mut w: Box<dyn IcebergWriter>) {
+    async fn _guarantee_object_safe(mut w: BoxedIcebergWriter) {
```

Comment:
We can add some case in this function to make sure type works in compile time and as the example of how can we use them🤔, like 
```
 async fn _guarantee_object_safe(
        w: impl IcebergWriter,
        w_b: impl IcebergWriterBuilder,
        mut box_w: BoxedIcebergWriter,
        box_w_b: BoxedIcebergWriterBuilder,
    ) {
        let _ = box_w
            .write(RecordBatch::new_empty(Schema::empty().into()))
            .await;
        let _ = box_w.close().await;
        let _box_w = box_w_b.build().await.unwrap();

        let box_w_b = w_b.clone().boxed();
        let _box_w = box_w_b.build().await.unwrap();
        let _box_w = w.boxed();
        let _box_w = w_b.build().await.unwrap().boxed();
    }
```